File: external/streamdeck/utils/generate-icons/ensure-cairo.mjs
// scripts/ensure-cairo.mjs
import { execSync } from "node:child_process";

function has(cmd) {
  try {
    execSync(cmd, { stdio: "ignore" });
    return true;
  } catch {
    return false;
  }
}

function hasCairoDll() {
  // names account for conda miniforge3 and Windows GTK installations
  return has("where.exe cairo.dll") || has("where.exe libcairo-2.dll");
}

const condaBin = `${process.env.USERPROFILE}\\miniforge3\\Library\\bin`;

if (!hasCairoDll()) {
  console.log("libcairo-2.dll not found.");

  if (!has("where.exe conda")) {
    console.warn(
      "conda not found. Install Miniforge first: " +
        "winget install --id CondaForge.Miniforge3 -e " +
        "then restart your shell and re-run `npm run setup:cairo`.",
    );
    process.exit(1);
  }

  console.log("Installing cairo via conda-forge...");
  execSync("conda install -c conda-forge cairo -y", { stdio: "inherit" });
  console.log(
    `Cairo installed. Ensure "${condaBin}" is on your PATH, then restart your shell.`,
  );
} else {
  console.log("libcairo already present.");
}

File: src/scripts/generate_streamdeck_icons/ensure_cairo.py
"""Ensure the cairo native library required by cairosvg is installed.

cairosvg needs ``cairo.dll`` / ``libcairo-2.dll`` on PATH.
"""

import os
import shutil
import subprocess
import sys
from pathlib import Path

CAIRO_DLL_NAMES = ("cairo.dll", "libcairo-2.dll")  # Windows GTK and conda names
CONDA_BIN = Path(os.environ.get("USERPROFILE", "")) / "miniforge3" / "Library" / "bin"


def cairo_dll_found() -> bool:
    path_dirs = os.environ.get("PATH", "").split(os.pathsep)
    return any(
        (Path(d) / name).is_file() for d in path_dirs if d for name in CAIRO_DLL_NAMES
    )


def ensure_cairo() -> int:
    if cairo_dll_found():
        print("libcairo already present.")
        return 0

    print("libcairo-2.dll not found.")
    if shutil.which("conda") is None:
        print(
            "conda not found. Install Miniforge first:\n"
            "  winget install --id CondaForge.Miniforge3 -e\n"
            "then restart your shell and re-run `sdeck-ensure-cairo`.",
            file=sys.stderr,
        )
        return 1

    print("Installing cairo via conda-forge...")
    subprocess.run(["conda", "install", "-c", "conda-forge", "cairo", "-y"], check=True)
    print(
        f'Cairo installed. Ensure "{CONDA_BIN}" is on your PATH, then restart your shell.'
    )
    return 0


def main() -> None:
    raise SystemExit(ensure_cairo())


if __name__ == "__main__":
    main()
